docs/proposals/lvmd-within-topolvm-node.md

```diff
@@ -0,0 +1,111 @@
+# Starting LVMD from within topolvm-node instead of using gRPC
+
+## Motivation
+
+To run TopoLVM on Edge currently requires quite a lot of memory for a minimal installation. One of the main
+reasons for that is due to the lvmd container running as a separate DaemonSet from the topolvm-node DaemonSet.
+The idea is to combine these two into a single DaemonSet to reduce the memory footprint as they are usually run
+within the same pod anyway when lvmd is not running as a systemd service.
+
+### Goals
+
+- Unify the topolvm-node and lvmd containers into a single container and not have to communicate via gRPC.
+- Allow lvmd to run as a systemd service or as a DaemonSet like before.
+- Do not break existing installations.
+- Reduce memory footprint.
+
+## Proposal
+
+TopoLVM is a storage plugin based on [CSI](https://github.com/container-storage-interface/spec/).
+Therefore, the architecture basically follows the one described in
+https://kubernetes-csi.github.io/docs/ .
+
+To manage LVM, `lvmd` should be run as a system service of the node OS.
+It provides gRPC services via UNIX domain socket to create/update/delete
+LVM logical volumes and watch a volume group status.
+
+`topolvm-node` implements CSI node services as well as miscellaneous control
+on each Node.  It communicates with `lvmd` to watch changes in free space
+of a volume group and exports the information by annotating Kubernetes
+`Node` resource of the running node.  In the meantime, it adds a finalizer
+to the `Node` to clean up PersistentVolumeClaims (PVC) bound on the node.
+
+Since both topolvm-node and lvmd need to run as DaemonSet, we can simply combine them and start them under a single binary.
+
+## Design Details
+
+### Allowing topolvm-node to start lvmd
+
+The main problem with running lvmd as a systemd service or DaemonSet is that it is not possible to start it from within the topolvm-node container.
+We can fix this by allowing topolvm-node to start lvmd by compiling a service into its startup command.
+This can be enabled via flag and then use a wrapper to call lvmd server commands directly from the client.
+
+### Flow of Operations for lvmd running within topolvm-node
+
+1. topolvm-node starts and checks if lvmd should be started as well via `--llvmd` flag.
```

Comment:
How about changing the option name to embed-lvmd? It's for the following two reasons.

- llvmd would be misunderstood as something for LLVM.
- Embed is more proper word than local in this case. It's because lvmd runs on local node in both the existing implementation and in the new mode.

LocalServiceClients will also be changed to EmbeddedServiceClients.

After reflecting this change request, I'll approve this PR.

Reply:
Great proposal for naming, adjusted accordingly and squashed the PR to prepare it for merge